File: game-ui-harnesses/ui-component-harness/src/select-menu-highlights.ts
import type {Issue} from './contract.ts';
export interface SelectMenuHighlight {
  color: string;
  alpha: number;
  insets: {top:number;right:number;bottom:number;left:number};
  cornerRadius: number;
}
/** Equal-height popup content rows; selected wins over hover, never stacked. */
export interface SelectMenuHighlights {
  version: '1.0';
  coordinateSpace: 'popup-row-local';
  selected: SelectMenuHighlight;
  hover: SelectMenuHighlight;
}
export function validateSelectMenuHighlights(value:unknown, content:unknown, count:number, path:string):Issue[]{
  const issues:Issue[]=[];
  const fail=(p:string,code:string,message:string)=>issues.push({path:p,code,message});
  const object=(v:unknown,p:string,keys:string[]):Record<string,unknown>|undefined=>{
    if(!v||typeof v!=='object'||Array.isArray(v)){fail(p,'OBJECT_REQUIRED','must be an object');return;}
    const o=v as Record<string,unknown>;
    for(const k of keys)if(!Object.hasOwn(o,k))fail(p+'.'+k,'REQUIRED_FIELD','field is required');
    for(const k of Object.keys(o))if(!keys.includes(k))fail(p+'.'+k,'UNKNOWN_FIELD','unsupported field');
    return o;
  };
  const data=object(value,path,['version','coordinateSpace','selected','hover']);if(!data)return issues;
  if(data.version!=='1.0')fail(path+'.version','UNSUPPORTED_VERSION','only 1.0 is supported');
  if(data.coordinateSpace!=='popup-row-local')fail(path+'.coordinateSpace','UNSUPPORTED_COORDINATE_SPACE','must be popup-row-local');
  const safe=content as {width:number;height:number}|undefined;
  const valid=!!safe&&[safe.width,safe.height,count].every(n=>Number.isFinite(n)&&n>0);
  if(!valid)fail(path,'POPUP_CONTENT_REQUIRED','menuHighlights requires explicit valid popupContentLayout and nonempty options');
  for(const name of ['selected','hover']){
    const p=path+'.'+name,state=object(data[name],p,['color','alpha','insets','cornerRadius']);if(!state)continue;
    if(typeof state.color!=='string'||!/^#[0-9a-fA-F]{6}$/.test(state.color))fail(p+'.color','INVALID_COLOR','must be #RRGGBB');
    if(typeof state.alpha!=='number'||!Number.isFinite(state.alpha)||state.alpha<0||state.alpha>1)fail(p+'.alpha','INVALID_ALPHA','must be finite in [0,1]');
    const inset=object(state.insets,p+'.insets',['top','right','bottom','left']);
    const nonnegative=(v:unknown)=>typeof v==='number'&&Number.isFinite(v)&&v>=0;
    if(!nonnegative(state.cornerRadius))fail(p+'.cornerRadius','INVALID_RADIUS','must be finite and nonnegative');
    if(inset){
      for(const key of ['top','right','bottom','left'])if(!nonnegative(inset[key]))fail(p+'.insets.'+key,'INVALID_INSET','must be finite and nonnegative');
      if(valid&&Object.values(inset).every(nonnegative)){
        const width=safe!.width-(inset.left as number)-(inset.right as number),height=safe!.height/count-(inset.top as number)-(inset.bottom as number);
        if(width<=0||height<=0)fail(p+'.insets','HIGHLIGHT_OUT_OF_BOUNDS','insets must leave a positive rectangle inside each content row');
        else if(nonnegative(state.cornerRadius)&&(state.cornerRadius as number)>Math.min(width,height)/2)fail(p+'.cornerRadius','RADIUS_OUT_OF_BOUNDS','radius must not exceed half of the smaller inset dimension');
      }
    }
  }
  return issues;
}
export function scaleSelectMenuHighlights(value:SelectMenuHighlights,scale:number):SelectMenuHighlights{
  const part=(v:SelectMenuHighlight):SelectMenuHighlight=>({...v,insets:{top:v.insets.top*scale,right:v.insets.right*scale,bottom:v.insets.bottom*scale,left:v.insets.left*scale},cornerRadius:v.cornerRadius*scale});
  return {...value,selected:part(value.selected),hover:part(value.hover)};
}

File: game-ui-harnesses/ui-decomposition-harness/src/ai_ui_decomposition/select_menu_highlights.py
"""Mirror the consumer's select-menu-highlights-v1; never infer a palette."""
import copy
import math
import re
from .common import require


def _object(value, keys, suffix):
    require(isinstance(value, dict) and set(value) == set(keys), 'SELECT_MENU_HIGHLIGHTS_' + suffix)


def _number(value):
    return type(value) in (int, float) and math.isfinite(value)


def validate_menu_highlights(state, option_count, popup_size=None, *, runtime=False):
    if 'menuHighlights' not in state:
        return None
    value = state['menuHighlights']
    _object(value, ('version','coordinateSpace','selected','hover'), 'FIELDS')
    require(value['version'] == '1.0', 'SELECT_MENU_HIGHLIGHTS_VERSION')
    require(value['coordinateSpace'] == 'popup-row-local', 'SELECT_MENU_HIGHLIGHTS_COORDINATE_SPACE')
    safe = state.get('popupContentLayout')
    _object(safe, ('x','y','width','height') if runtime else ('coordinateSpace','x','y','width','height'), 'POPUP_CONTENT_REQUIRED')
    require(runtime or safe['coordinateSpace'] == 'target-popup-local', 'SELECT_MENU_HIGHLIGHTS_COORDINATE_SPACE')
    require(all(_number(safe[k]) for k in ('x','y','width','height')) and safe['x'] >= 0 and safe['y'] >= 0 and safe['width'] > 0 and safe['height'] > 0, 'SELECT_MENU_HIGHLIGHTS_POPUP_GEOMETRY')
    require(type(option_count) is int and option_count > 0, 'SELECT_MENU_HIGHLIGHTS_OPTIONS')
    if popup_size is not None:
        require(safe['x'] + safe['width'] <= popup_size[0] and safe['y'] + safe['height'] <= popup_size[1], 'SELECT_MENU_HIGHLIGHTS_POPUP_GEOMETRY')
    for key in ('selected','hover'):
        part = value[key]
        _object(part, ('color','alpha','insets','cornerRadius'), 'STATE_FIELDS')
        require(isinstance(part['color'], str) and re.fullmatch(r'#[0-9a-fA-F]{6}', part['color']), 'SELECT_MENU_HIGHLIGHTS_COLOR')
        require(_number(part['alpha']) and 0 <= part['alpha'] <= 1, 'SELECT_MENU_HIGHLIGHTS_ALPHA')
        _object(part['insets'], ('top','right','bottom','left'), 'INSETS_FIELDS')
        inset = part['insets']; radius = part['cornerRadius']
        require(all(_number(v) and v >= 0 for v in inset.values()), 'SELECT_MENU_HIGHLIGHTS_INSET')
        require(_number(radius) and radius >= 0, 'SELECT_MENU_HIGHLIGHTS_RADIUS')
        width = safe['width'] - inset['left'] - inset['right']
        height = safe['height'] / option_count - inset['top'] - inset['bottom']
        require(width > 0 and height > 0, 'SELECT_MENU_HIGHLIGHTS_ROW_GEOMETRY')
        require(radius <= min(width,height)/2, 'SELECT_MENU_HIGHLIGHTS_RADIUS_GEOMETRY')
    return value


def scale_menu_highlights(value, scale):
    result = copy.deepcopy(value)
    for key in ('selected','hover'):
        result[key]['insets'] = {k:v*scale for k,v in value[key]['insets'].items()}
        result[key]['cornerRadius'] *= scale
    return result


def require_runtime_menu_highlights(state, appearance, option_count, registration_scale):
    authored = validate_menu_highlights(state, option_count)
    actual = validate_menu_highlights(appearance, option_count, [appearance['popupCanvas']['width'],appearance['popupCanvas']['height']], runtime=True)
    require((authored is None) == (actual is None), 'STATE_SELECT_MENU_HIGHLIGHTS_MISMATCH')
    if authored is not None:
        expected = scale_menu_highlights(authored, 1/registration_scale)
        require(expected == actual, 'STATE_SELECT_MENU_HIGHLIGHTS_MISMATCH')
    return actual
